### DISCUSSION TITLE
How to map field name when using `$queryRaw`?

### DISCUSSION THREAD

**Discussion by wongsean** [prisma/prisma#12324] [Q&A] [answered]
Hello everyone, I have a problem with field name mapping while using `$queryRaw`.
For example, this is the schema:
```
model User {
  id       String @id
  lastName String @map("last_name")

  @@map("users")
}

// types generated
type User = {
  id: string
  lastName: string
}

await prisma.$queryRaw`select * from users where id = ${id}`
// => { id: string, last_name: string }[]

await prisma.$queryRaw`select id, last_name as lastName from users where id = ${id}`
// => { id: string, lastname: string }[]

// what I want is User[]
```

Is there a way to map field names like "last_name" to "lastName" when using `$queryRaw`?
Currently, my workaround is writing a `MapKeysToSnakeCase` type helper and annotating the query results as `MapKeysToSnakeCase<User>[]`, but it's not likely to be compatible with regular handling-user logic. I can also do fields mapping by myself, but it sort of breaks the principle of "single source of truth".

So, I am wondering if there is a way out of boxes that I can map raw query results to the model I defined?

**Comment by austincrim** [comment, marked as answer]
Hey @wongsean 👋 ,

There is not a built-in way to do this (unless you want to remove the `@map` in your schema 😛), but you could potentially write [Prisma Middleware](https://www.prisma.io/docs/concepts/components/prisma-client/middleware) to do this.

This might be a starting point:
```ts
const prisma = new PrismaClient()

prisma.$use(async (params, next) => {
  let results = await next(params)
  if (params.action === 'queryRaw') {
    results.forEach((r) => {
      r['lastName'] = r['last_name']
      delete r.last_name
    })
  }

  return results
})
```

If you have an idea for how we could solve this in Prisma, please open a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=)!

**Reply by wongsean** [reply]
Hi, @austincrim thanks!

It's a brand new service, so I will remove `@map` to avoid the mapping stuff here😄.

**Comment by ellisza** [comment]
For anyone else who ends up at this question an alternative is to cast the column names in the rawQuery. The following is an example from a similar issue I had and my solution.

`prisma.$queryRaw'SELECT uuid, full_name AS "fullName", birth_date AS "birthDate"`

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
